lib/openstack_query/query_server.py

```diff
@@ -0,0 +1,144 @@
+            ServerProperties.IMAGE_ID: lambda **kwargs: {"image": kwargs["value"]},
+            ServerProperties.PROJECT_ID: lambda **kwargs: {
+                "project_id": kwargs["value"]
+            },
+        },
+        QueryPresets.OLDER_THAN_OR_EQUAL_TO: {
+            ServerProperties.SERVER_LAST_UPDATED_DATE: lambda **kwargs: {
+                "changes-before": convert_to_timestamp(**kwargs)
+            }
+        },
+        QueryPresets.YOUNGER_THAN_OR_EQUAL_TO: {
+            ServerProperties.SERVER_LAST_UPDATED_DATE: lambda **kwargs: {
+                "changes-since": convert_to_timestamp(**kwargs)
+            }
+        },
+    }
+
+    _DEFAULT_FILTER_FUNCTION_MAPPINGS = {
+        QueryPresets.EQUAL_TO: ["*"],
+        QueryPresets.NOT_EQUAL_TO: ["*"],
+        QueryPresets.OLDER_THAN: [
+            ServerProperties.SERVER_CREATION_DATE,
+            ServerProperties.SERVER_LAST_UPDATED_DATE,
+        ],
+        QueryPresets.YOUNGER_THAN: [
+            ServerProperties.SERVER_CREATION_DATE,
+            ServerProperties.SERVER_LAST_UPDATED_DATE,
+        ],
+        QueryPresets.YOUNGER_THAN_OR_EQUAL_TO: [
+            ServerProperties.SERVER_CREATION_DATE,
+            ServerProperties.SERVER_LAST_UPDATED_DATE,
+        ],
+        QueryPresets.OLDER_THAN_OR_EQUAL_TO: [
+            ServerProperties.SERVER_CREATION_DATE,
+            ServerProperties.SERVER_LAST_UPDATED_DATE,
+        ],
+        QueryPresets.MATCHES_REGEX: [ServerProperties.SERVER_NAME],
+    }
+
+    _NON_DEFAULT_FILTER_FUNCTION_MAPPINGS = {}
+
+    def __init__(self, connection_cls=OpenstackConnection):
+        QueryWrapper.__init__(self, connection_cls)
+
+    @staticmethod
+    def _run_query(
+        conn: OpenstackConnection,
+        filter_kwargs: Optional[Dict[str, str]] = None,
+        **kwargs
+    ):
+        """
+        This method runs the query by running openstacksdk commands
+        For QueryServer, this command gets all projects available and iteratively finds servers in that project
+        :param conn: An OpenstackConnection object
+        :param filter_kwargs: An Optional set of kwargs to pass to conn.compute.servers
+        :param kwargs: a set of extra meta params that override openstacksdk commands or change which commands are run
+
+        kwargs: For QueryServer:
+            'from_projects' - set which projects to search by providing a list of openstack projects to search in
+        """
+        servers = []
+        if "from_projects" in kwargs:
+            projects = kwargs["from_projects"]
+        else:
+            projects = []
+            for project in conn.identity.projects():
+                projects.append(project)
+
+        for project in projects:
+            servers.extend(
+                QueryServer._run_query_on_project(conn, project, filter_kwargs)
+            )
+        return servers
+
+    @staticmethod
+    def _run_query_on_project(
```

Comment:
Should this be called something like `_run_query_on_all_projects`, since the `all_tenants` implies that to an extent?
This would only be runnable by admins because of the all_tenants flag

Reply:
no we're filtering by project_id - the `all_tenants` forces the search context to search in all projects (just in case that's needed) - this was copied from the previous method that was used